File: src/pages/Jobs.ts
import { Browser, ElementHandle, Page } from "puppeteer";
import BasePage from "./Base";
import { JobDetails, type JobFilter } from "../../typings/index";
import config from "../../config.js";
import Utils from "../utils";

class JobsPage extends BasePage {
    #path = "/jobs/search";
    #technologiesByYears: Map<string, number>;

    constructor(browser: Browser, page: Page) {
        super(browser, page);
        this.#technologiesByYears = new Map<string, number>();

        const expConfig: { [key: number]: string[] } = config.answers.experience_years;
        for (const year of Object.keys(expConfig)) {
            const yearAsNumber = parseInt(year, 10);
            const technologies = expConfig[yearAsNumber];
            for (const technology of technologies) {
                this.#technologiesByYears.set(technology, parseInt(year));
            }
        }
    }

    async open() {
        if (!this.page) {
            this.page = super.page = await this.browser.newPage();
        }
        await this.page.goto(this.baseUrl + this.#path);
    }

    async applyFilters(filters: JobFilter) {
        const { title, location, easyApply, workLocation, jobType, experienceLevel } = filters;
        let url = this.baseUrl + this.#path;

        const params = new URLSearchParams();
        if (title) params.append("keywords", title);
        if (location) params.append("location", location);
        if (easyApply) params.append("f_AL", "true");

        if (workLocation) {
            if (workLocation === "ON_SITE") params.append("f_WT", "1");
            else if (workLocation === "REMOTE") params.append("f_WT", "2");
            else if (workLocation === "HYBRID") params.append("f_WT", "3");
        }

        if (jobType) {
            if (jobType === "PART_TIME") params.append("f_JT", "P");
            else if (jobType === "FULL_TIME") params.append("f_JT", "F");
            else if (jobType === "CONTRACT") params.append("f_JT", "C");
            else if (jobType === "TEMPORARY") params.append("f_JT", "T");
            else if (jobType === "OTHER") params.append("f_JT", "O");
        }

        if (experienceLevel) {
            if (experienceLevel === "INTERNSHIP") params.append("f_E", "1");
            else if (experienceLevel === "ENTRY_LEVEL") params.append("f_E", "2");
            else if (experienceLevel === "ASSOCIATE") params.append("f_E", "3");
            else if (experienceLevel === "MID_SENIOR") params.append("f_E", "4");
            else if (experienceLevel === "DIRECTOR") params.append("f_E", "5");
            else if (experienceLevel === "EXECUTIVE") params.append("f_E", "6");
        }

        url += "?" + params.toString();

        // open URL
        await this.page.goto(url);

        // wait for results to load
        const el = await this.page.waitForSelector("small.jobs-search-results-list__text");
        await this.wait(5000);

        // get results
        const count = await this.page.evaluate((el) => el?.innerText, el);

        // log results
        return count;
    }

    async apply() {
        const ul = await this.page.$(
            ".jobs-search-results-list > ul.scaffold-layout__list-container"
        );

        if (!ul) {
            return console.warn("Jobs Results Missing on this page!");
        }

        // TODO: Pagination logic

        const pageJobs = await ul.$$("li");
        for (let i = 1; i <= pageJobs.length; i++) {
            // wait and click i'th element
            await this.targetClick(
                `.jobs-search-results-list > ul.scaffold-layout__list-container > li:nth-child(${i})`
            );

            await this.wait(5000);

            // Log job details
            const jobDetails = await this.#parseJobDetails().catch((err) => console.log(err));
            if (!jobDetails) {
                console.log("Failed to parse job details");
                continue;
            }

            let closeModal = false;
            // Try Apply
            try {
                await this.#easyApply(jobDetails);
                if (config.testMode) {
                    closeModal = true; // close modal since we are not actually applying
                    console.log("[TEST MODE]: Applied to this job");
                } else {
                    console.log("Applied to this job");
                }
                Utils.recordApplicationStatus(jobDetails, "SUCCESS");
            } catch (err: any) {
                console.log(`Failed to apply to this job. Reason: ${err.message}`);
                Utils.recordApplicationStatus(jobDetails, "FAILED", err.message);
                closeModal = true;
            }

            if (closeModal) {
                await this.targetClick(".artdeco-modal__dismiss");
                await this.wait(2000);
                await this.targetClick(".artdeco-modal__confirm-dialog-btn:nth-child(1)");
            }
        }
    }

    async #easyApply(jobDetails: JobDetails) {
        // EasyApply button
        const btn = await this.page.$("div.jobs-details__main-content button.jobs-apply-button");
        if (!btn) {
            throw new Error("No Easy Apply button found");
        }
        await btn.click();

        await this.wait(5000);

        // track progress
        const progressEl = await this.page.$(
            ".jobs-easy-apply-content progress.artdeco-completeness-meter-linear__progress-element"
        );

        if (!progressEl) {
            console.log("No progress element found! Considering 100% progress");
        }

        let progress = progressEl ? await progressEl.evaluate((el) => el?.value) : 100;

        while (progress < 100) {
            await this.wait(5000);

            let button = null as ElementHandle<Element> | null;

            // next button
            button = await this.page.$(
                '.jobs-easy-apply-content footer button[aria-label="Continue to next step"]'
            );

            // review button
            if (!button) {
                button = await this.page.$(
                    '.jobs-easy-apply-content footer button[aria-label="Review your application"]'
                );
            }

            if (!button) {
                throw new Error("No Next/Review button found");
            }

            await this.#tryFillDetails(jobDetails);

            await button.click();
            await this.wait(2000);
            const newProgress = await this.page.evaluate((el) => el?.value, progressEl);
            if (!newProgress) {
                // TODO: Check if this is the last step
                throw new Error("No progress found");
            }

            if (newProgress === progress) {
                throw new Error("Could not fill details");
            }

            progress = newProgress;
        }

        // Don't submit in test mode
        if (config.testMode) {
            return;
        }

        // Submit button
        await this.wait(5000);
        const submitBtn = await this.page.$(
            '.jobs-easy-apply-content footer button[aria-label="Submit application"]'
        );
        if (submitBtn) {
            await submitBtn.click();
        }
    }

    async #tryFillDetails(jobDetails: JobDetails) {
        const pb4 = await this.page.$$(".jobs-easy-apply-content div.pb4");

        if (pb4.length === 0) {
            console.log("No b4 element found");
            return;
        }

        for (let i = 0; i < pb4.length; i++) {
            const b4 = pb4[i];
            try {
                await this.#tryUploadResume(b4);
            } catch (err) {}

            try {
                await this.#tryFillQuestions(b4, jobDetails);
            } catch (err) {}
        }
    }

    async #tryUploadResume(b4: ElementHandle<HTMLDivElement>) {
        const input = await b4.$("input[name='file']");
        if (!input) return;

        const span = await b4.$("span[role='button']");
        const innerText = await span?.evaluate((el) => el?.innerText.toLowerCase());

        // Resume
        if (innerText?.includes("resume")) {
            const checkedResume = await b4.$("input[type='radio']:checked");
            if (checkedResume) return;
            const uploadedResumes = await b4.$$("input[type='radio']");

            // Already uploaded
            if (uploadedResumes.length > 0) {
                const latestResume = await b4.$("input[type='radio'] + label");
                if (latestResume) {
                    await latestResume.click();
                    return;
                }
            }

            // TODO: Upload resume
            else {
            }
        }

        // Cover Letter
        if (innerText?.includes("cover")) {
            // TODO: Cover Letter Logic
        }
    }

    async #tryFillQuestions(b4: ElementHandle<HTMLDivElement>, jobDetails: JobDetails) {
        const frmEls = await b4.$$("div.jobs-easy-apply-form-section__grouping");
        let done;

        for (let i = 0; i < frmEls.length; i++) {
            const el = frmEls[i];
            try {
                done = await this.#tryFillRadio(el, jobDetails);
            } catch (err) {
                console.log("Failed to fill radio", err);
            }

            if (done) continue;

            try {
                done = await this.#tryFillDropdown(el, jobDetails);
            } catch (err) {
                console.log("Failed to fill dropdown", err);
            }

            if (done) continue;

            try {
                await this.#tryFillTextBox(el, jobDetails);
            } catch (err) {
                console.log("Failed to fill text box", err);
            }
        }
    }

    async #tryFillTextBox(el: ElementHandle<Element>, jobDetails: JobDetails) {
        const quesEl = await el.$("label");
        const txtField = await el.$("input");
        const txtArea = await el.$("textarea");
        if (!quesEl || (!txtField && !txtArea)) return false;
        const ques = await quesEl.evaluate((el) => el.innerText.toLowerCase().trim());

        // check if text box is already filled
        const value = txtField
            ? await txtField?.evaluate((el) => el.value)
            : await txtArea?.evaluate((el) => el.value);

        const answered = typeof value === "string" && value.length > 0;

        let toEnter = null;
        if (ques.includes("experience")) {
            let years = null;
            for (const [tech, y] of this.#technologiesByYears) {
                if (ques.includes(tech.toLowerCase())) {
                    years = y.toString();
                    break;
                }
            }
            if (!years) {
                years = config.answers.default_experience_years.toString();
            }
            toEnter = years;
        } else if (ques.includes("first name")) {
            toEnter = config.answers.first_name;
        } else if (ques.includes("last name")) {
            toEnter = config.answers.last_name;
        } else if (ques.includes("name")) {
            toEnter = config.answers.first_name + " " + config.answers.last_name;
        } else if (ques.includes("mobile phone")) {
            toEnter = config.answers.phone;
        } else if (ques.includes("website") || ques.includes("portfolio")) {
            toEnter = config.answers.website;
        } else if (ques.includes("linkedin")) {
            toEnter = config.answers.linkedin;
        } else if (ques.includes("github")) {
            toEnter = config.answers.github;
        } else if (ques.includes("street")) {
            toEnter = config.answers.street_address;
        } else if (ques.includes("city")) {
            toEnter = config.answers.city;
        } else if (ques.includes("state") || ques.includes("province")) {
            toEnter = config.answers.state;
        } else if (ques.includes("zip") || ques.includes("postal")) {
            toEnter = config.answers.zip_code;
        }

        // TODO: Fill remaining text boxes

        if (toEnter !== null) {
            if (!answered) {
                await txtField?.type(toEnter, { delay: 1000 });
                await txtArea?.type(toEnter, { delay: 1000 });
            }
            Utils.recordPreparedQuestion(jobDetails, "TEXT", ques, toEnter);
        } else {
            Utils.recordUnpreparedQuestion(jobDetails, "TEXT", ques);
        }

        return true;
    }

    async #tryFillRadio(el: ElementHandle<Element>, jobDetails: JobDetails) {
        const quesEl = await el.$("div.jobs-easy-apply-form-element legend");
        const radios = await el.$$(".fb-text-selectable__option input");
        if (!quesEl || radios.length === 0) return false;
        const ques = await quesEl.evaluate((el) => el.innerText.toLowerCase().trim());

        // check if any radio is already selected
        let answered = false;
        for (const radio of radios) {
            const checked = await radio.evaluate((el) => el.checked);
            if (checked) {
                answered = true;
                break;
            }
        }

        // TODO: Fill radio
        if (answered) {
            Utils.recordPreparedQuestion(jobDetails, "RADIO", ques, "TODO");
        } else {
            Utils.recordUnpreparedQuestion(jobDetails, "RADIO", ques);
        }
        return true;
    }

    async #tryFillDropdown(el: ElementHandle<Element>, jobDetails: JobDetails) {
        const quesEl = await el.$("div.jobs-easy-apply-form-element");
        const select = await quesEl?.$("select");
        if (!quesEl || !select) return false;
        const quesLabel = await quesEl.$("label");
        if (!quesLabel) return false;
        const ques = await quesLabel.evaluate((el) => el.innerText.toLowerCase().trim());

        // check if an option is already selected
        const filled = await select.evaluate((el) => el.value);
        const answered = typeof filled === "string" && filled.length > 0;

        if (answered) {
            Utils.recordPreparedQuestion(jobDetails, "DROPDOWN", ques, filled);
        } else {
            Utils.recordUnpreparedQuestion(jobDetails, "DROPDOWN", ques);
        }

        return true;
    }

    async #parseJobDetails() {
        let sel = ".jobs-details__main-content .jobs-unified-top-card";
        const title = await this.page.$eval(
            `${sel} h2.jobs-unified-top-card__job-title`,
            (el) => el.innerText
        );

        const company = await this.page.$eval(
            `${sel} span.jobs-unified-top-card__company-name`,
            (el) => el.innerText
        );

        const location = await this.page.$eval(
            `${sel} span.jobs-unified-top-card__bullet`,
            (el) => el.innerText
        );

        const workType = await this.page.$eval(
            `${sel} span.jobs-unified-top-card__workplace-type`,
            (el) => el.innerText
        );

        const applicants = await this.page.$eval(
            `${sel} .jobs-unified-top-card__subtitle-secondary-grouping > span.jobs-unified-top-card__bullet`,
            (el) => el.innerText
        );

        const posted = await this.page.$eval(
            `${sel} span.jobs-unified-top-card__posted-date`,
            (el) => el.innerText
        );

        const jobId = this.page.url().split("currentJobId=")[1].split("&")[0];

        // TODO: Style this
        console.log("====================================");
        console.log("Title: ", title);
        console.log("Company: ", company);
        console.log("Location: ", location);
        console.log("Work Type: ", workType);
        console.log("Applicants: ", applicants);
        console.log("Posted: ", posted);

        return { jobId, title, company, location, workType, applicants, posted };
    }
}

export default JobsPage;

File: src/pages/Jobs.py
from selenium.webdriver.common.by import By
from selenium.webdriver.remote.webelement import WebElement
from .Base import BasePage
from urllib.parse import urlencode, urlparse, parse_qs
from providers.answer_providers import AnswerProvider
from utils import (
    config,
    record_answered_question,
    record_unprepared_question,
    record_application_status,
)
from typing import Literal


class JobDetails:
    def __init__(self, url: str, el: WebElement) -> None:
        self.url = url
        self.job_id = parse_qs(urlparse(url).query).get("currentJobId").pop()

        try:
            self.title = el.find_element(
                By.CSS_SELECTOR,
                ".jobs-unified-top-card span.job-details-jobs-unified-top-card__job-title-link",
            ).text
        except Exception:
            self.title = ""

        try:
            details_el = el.find_element(
                By.CSS_SELECTOR,
                ".jobs-details__main-content .jobs-unified-top-card .job-details-jobs-unified-top-card__primary-description-without-tagline",
            ).text
            details = details_el.split("·")
            self.company = details[0].strip()
            self.location = details[1].strip()
            self.posted = details[2].strip()
            self.applicants = details[3].strip()
        except Exception:
            self.company = self.location = self.posted = self.applicants = ""

        print(
            f"""
        JobId: {self.job_id}
        Title: {self.title}
        Company: {self.company}
        Location: {self.location}
        Applicants: {self.applicants}
        Posted: {self.posted}
        """
        )


class JobsPage(BasePage):
    path = "/jobs/search"

    def __init__(self, driver, answer_provider: AnswerProvider) -> None:
        super().__init__(driver)
        self.current_job = None
        self.answer_provider = answer_provider

    def open(self):
        self.driver.get(self.base_url + self.path)
        return self

    def apply_filters(
        self,
        title=None,
        location=None,
        work_location: Literal["ON_SITE", "REMOTE", "HYBRID"] = None,
        job_type: Literal["PART_TIME", "FULL_TIME", "CONTRACT", "TEMPORARY", "OTHER"] = None,
        experience_level: Literal[
            "INTERNSHIP",
            "ENTRY_LEVEL",
            "ASSOCIATE",
            "MID_SENIOR",
            "DIRECTOR",
            "EXECUTIVE",
        ] = None,
    ):

        params = {
            "f_AL": "true",  # Easy Apply
        }

        if title:
            params["keywords"] = title

        if location:
            params["location"] = location

        if work_location:
            if work_location == "ON_SITE":
                params["f_WT"] = "1"
            elif work_location == "REMOTE":
                params["f_WT"] = "2"
            elif work_location == "HYBRID":
                params["f_WT"] = "3"

        if job_type:
            if job_type == "PART_TIME":
                params["f_JT"] = "P"
            elif job_type == "FULL_TIME":
                params["f_JT"] = "F"
            elif job_type == "CONTRACT":
                params["f_JT"] = "C"
            elif job_type == "TEMPORARY":
                params["f_JT"] = "T"
            elif job_type == "OTHER":
                params["f_JT"] = "O"

        if experience_level:
            if experience_level == "INTERNSHIP":
                params["f_E"] = "1"
            elif experience_level == "ENTRY_LEVEL":
                params["f_E"] = "2"
            elif experience_level == "ASSOCIATE":
                params["f_E"] = "3"
            elif experience_level == "MID_SENIOR":
                params["f_E"] = "4"
            elif experience_level == "DIRECTOR":
                params["f_E"] = "5"
            elif experience_level == "EXECUTIVE":
                params["f_E"] = "6"

        url = self.base_url + self.path + "?" + urlencode(params)

        # open URL
        self.driver.get(url)

        # wait for results to load
        el = self.wait_for_selector(By.CSS_SELECTOR, "small.jobs-search-results-list__text")

        print("Total jobs found:", el.text)

    def apply(self):
        ul = self.wait_for_selector(
            By.CSS_SELECTOR,
            ".jobs-search-results-list > ul.scaffold-layout__list-container",
            fail=False,
        )

        if not ul:
            print("Jobs Results Missing on this page!")
            return

        # TODO: Pagination Logic

        page_jobs = ul.find_elements(By.CSS_SELECTOR, "li")
        for i in range(1, len(page_jobs) + 1):
            print(f"Applying to job {i} of {len(page_jobs)}")
            self.wait_and_click(
                By.CSS_SELECTOR,
                f".jobs-search-results-list > ul.scaffold-layout__list-container > li:nth-child({i})",
            )
            self.wait(1)
            self.current_job = JobDetails(self.driver.current_url, self.driver)

            applied = self.find_element(
                by=By.CSS_SELECTOR,
                value="li-icon.artdeco-inline-feedback__icon",
                fail=False,
            )

            if applied:
                print("Already applied to this job!")
                record_application_status(self.current_job, "ALREADY_APPLIED")
            else:
                self.easy_apply()
                self.wait(1)

    def easy_apply(self):
        failed = False
        started = False
        try:
            # EasyApply button
            btn = self.wait_for_selector(
                By.CSS_SELECTOR,
                "div.jobs-details__main-content button.jobs-apply-button",
                fail=False,
            )

            assert btn, "Easy Apply button not found!"
            btn.click()
            started = True

            # Job Search Safety Modal
            modal = self.wait_for_selector(
                By.CSS_SELECTOR,
                "div.job-details-pre-apply-safety-tips-modal__footer",
                timeout=2,
                fail=False,
            )

            if modal:
                modal.find_element(By.CSS_SELECTOR, "button.jobs-apply-button").click()

            # track progress
            progress_el = self.wait_for_selector(
                By.CSS_SELECTOR,
                ".jobs-easy-apply-content progress.artdeco-completeness-meter-linear__progress-element",
                fail=False,
            )

            progress = 100
            if progress_el:
                progress = int(progress_el.get_attribute("value"))
            else:
                print("No progress element found! Considering 100% progress")

            while progress < 100:
                button = None

                # next button
                button = self.wait_for_selector(
                    By.CSS_SELECTOR,
                    ".jobs-easy-apply-content footer button[aria-label='Continue to next step']",
                    fail=False,
                )

                # review button
                if button is None:
                    button = self.wait_for_selector(
                        By.CSS_SELECTOR,
                        ".jobs-easy-apply-content footer button[aria-label='Review your application']",
                        fail=False,
                    )

                assert button, "No next or review button found!"

                # Fill out the form
                pb4 = self.driver.find_elements(By.CSS_SELECTOR, ".jobs-easy-apply-content div.pb4")

                for el in pb4:
                    try:
                        self.try_upload_resume(el)
                    except Exception:
                        pass

                    try:
                        self.try_fill_questions(el)
                    except Exception:
                        pass

                button.click()
                self.wait(0.5)

                new_progress = self.wait_for_selector(
                    By.CSS_SELECTOR,
                    ".jobs-easy-apply-content progress.artdeco-completeness-meter-linear__progress-element",
                    fail=False,
                )

                assert new_progress, "Progress element not found!"
                new_progress = int(progress_el.get_attribute("value"))

                assert new_progress != progress, "Progress not updated! Could not fill details?"

                progress = new_progress

        except Exception as e:
            failed = True
            record_application_status(self.current_job, "FAILED", str(e))
            print(f"Failed to apply to this job. Reason: {e}")

        close_modal = True if failed and started else False

        if not failed:
            # Don't submit in test mode
            if config["dev"]["test_mode"]:
                print("Test mode. Not submitting application.")
                record_application_status(self.current_job, "APPLIED_TEST_MODE")
                close_modal = True
            else:
                self.wait_and_click(
                    By.CSS_SELECTOR,
                    "button.jobs-apply-form__submit-button",
                )
                record_application_status(self.current_job, "APPLIED")

        if close_modal:
            self.wait_and_click(
                By.CSS_SELECTOR,
                ".artdeco-modal__dismiss",
            )
            self.wait(0.5)
            self.wait_and_click(
                By.CSS_SELECTOR,
                ".artdeco-modal__confirm-dialog-btn:nth-child(1)",
            )

    def try_upload_resume(self, el: WebElement):
        inp = el.find_element(By.CSS_SELECTOR, "input[type='file']")
        if inp is None:
            return

        span = el.find_element(By.CSS_SELECTOR, "span[role='button']")
        inner_text = span.text.lower()

        # Resume
        if "resume" in inner_text:
            checked_resume = el.find_element(By.CSS_SELECTOR, "input[type='radio']:checked")
            if checked_resume:
                return
            uploaded_resumes = el.find_elements(By.CSS_SELECTOR, "input[type='radio'] + label")
            if uploaded_resumes:
                latest_resume = el.find_element("input[type='radio'] + label")
                if latest_resume:
                    latest_resume.click()
                    return

            else:
                pass  # TODO: Upload resume

        # Cover Letter
        if "cover letter" in inner_text:
            pass  # TODO: Cover Letter Logic

    def try_fill_questions(self, el: WebElement):
        form_els = el.find_elements(By.CSS_SELECTOR, "div.jobs-easy-apply-form-section__grouping")

        for el in form_els:
            done = False
            if not done:
                try:
                    done = self.try_fill_radio(el)
                except Exception:
                    print("Failed to fill radio")

            if not done:
                try:
                    done = self.try_fill_dropdown(el)
                except Exception:
                    print("Failed to fill dropdown")

            if not done:
                try:
                    done = self.try_fill_input(el)
                except Exception:
                    print("Failed to fill input")

    def try_fill_input(self, el: WebElement):
        ques_el = self.find_element(el, By.CSS_SELECTOR, "label", False)

        # TODO: Verify this logic
        if ques_el and self.find_element(ques_el, By.CSS_SELECTOR, "span[aria-hidden='true']", False) is not None:
            ques_el = self.find_element(ques_el, By.CSS_SELECTOR, "span[aria-hidden='true']", False)

        # Some don't have labels but have a preceding span
        if not ques_el:
            ques_el = self.find_element(
                el,
                By.XPATH,
                "preceding::span[contains(@class, 'jobs-easy-apply-form-section__group-title')][1]",
                False,
            )

        text_field = self.find_element(el, By.CSS_SELECTOR, "input", False)
        text_area = self.find_element(el, By.CSS_SELECTOR, "textarea", False)
        if not ques_el or (not text_field and not text_area):
            return False
        ques = ques_el.text.lower()

        value = text_field.get_attribute("value") if text_field else text_area.get_attribute("value")
        answered = value != ""

        if answered:
            record_answered_question(self.current_job, ques, "TEXT", value)
        else:
            to_enter = None
            if "years of experience" in ques or "work experience" in ques:
                for experience in self.answer_provider.get_basic_answer("experience"):
                    year = experience.get("years")
                    skills = experience.get("skills")
                    for tech in skills:
                        if tech in ques:
                            to_enter = year
                            break

                if not to_enter:
                    to_enter = self.answer_provider.get_basic_answer("default_experience")

            elif ques == "first name":
                to_enter = self.answer_provider.get_basic_answer("first_name")
            elif ques == "last name":
                to_enter = self.answer_provider.get_basic_answer("last_name")
            elif ques == "mobile phone number":
                to_enter = self.answer_provider.get_basic_answer("mobile_phone_number")
            elif ques == "email address":
                to_enter = self.answer_provider.get_basic_answer("email_address")
            elif ques == "city":
                to_enter = self.answer_provider.get_basic_answer("city")

            # Additional Questions
            if not to_enter:
                to_enter = self.answer_provider.get_answer(ques)

            # Enter the value
            if to_enter:
                if text_field:
                    text_field.send_keys(to_enter)
                else:
                    text_area.send_keys(to_enter)
                record_answered_question(self.current_job, ques, "TEXT", to_enter)
            else:
                record_unprepared_question(self.current_job, ques, "TEXT")

        return True

    def try_fill_radio(self, el: WebElement):
        ques_el = self.find_element(
            el, By.CSS_SELECTOR, "div.jobs-easy-apply-form-element legend span[aria-hidden='true']", False
        )
        radios = self.find_elements(el, By.CSS_SELECTOR, ".fb-text-selectable__option input", False)
        if not ques_el or len(radios) == 0:
            return False
        ques = ques_el.text.lower()

        answered = False
        options = []
        for radio in radios:
            options.append(radio.get_attribute("value"))
            if radio.is_selected():
                answered = radio.get_attribute("value")

        if answered:
            # TODO: Get already answered value
            record_answered_question(self.current_job, ques, "RADIO", answered)
        else:
            # TODO: Implement logic to select a radio
            record_unprepared_question(self.current_job, ques, "RADIO", options)

        return True

    def try_fill_dropdown(self, el: WebElement):
        ques_el = self.find_element(el, By.CSS_SELECTOR, "div.jobs-easy-apply-form-element", False)
        dropdown = self.find_element(el, By.CSS_SELECTOR, "select", False)
        if not ques_el or not dropdown:
            return False
        ques_label = self.find_element(ques_el, By.CSS_SELECTOR, "label span:not(.visually-hidden)", False)

        # Some don't have labels but have a preceding span
        if not ques_label:
            ques_label = self.find_element(
                el,
                By.XPATH,
                "preceding::span[contains(@class, 'jobs-easy-apply-form-section__group-title')][1]",
                False,
            )
        ques = ques_label.text.lower()
        answered = dropdown.get_attribute("value") not in ["", "Select an option"]

        options = []
        for option in dropdown.find_elements(By.CSS_SELECTOR, "option"):
            if option.get_attribute("value") not in ["", "Select an option"]:
                options.append(option.text)

        if answered:
            # TODO: Get already answered value
            record_answered_question(self.current_job, ques, "DROPDOWN", dropdown.get_attribute("value"))
        else:
            # TODO: Implement logic to select a dropdown
            record_unprepared_question(self.current_job, ques, "DROPDOWN", options)
        return True
